### PR TITLE
Make ANTLR 4 runtime maven artifacts OSGi-ready

### PR DESCRIPTION
**Pull request by deklanowski**
Added maven-bundle-plugin + metadata configuration to antlr4-runtime and
antlr4-annotations jar manifests to allow antlr4 runtime to be used to
execute parsers in OSGi runtime.

The addition of OSGi metadata as generated by the maven-bundle-plugin (actually it is generated by a tool called [Bnd](http://www.aqute.biz/Code/Bnd)  which is invoked by the plugin) adds to the manifest of *antlr4-runtime* and *antlr4-annotations* manifests as follows (standard jar manifest entries remain and are highlighted in bold):

Firstly antlr4-runtime:

__Manifest-Version: 1.0__
Bnd-LastModified: 1409190229833
__Build-Jdk: 1.7.0_25__
__Built-By: cd01__
Bundle-Description: The ANTLR 4 Runtime
Bundle-DocURL: http://www.antlr.org
Bundle-License: http://www.antlr.org/license.html
Bundle-ManifestVersion: 2
Bundle-Name: ANTLR 4 Runtime
Bundle-SymbolicName: org.antlr.antlr4-runtime
Bundle-Vendor: ANTLR
Bundle-Version: 4.4.0.SNAPSHOT
__Created-By: Apache Maven Bundle Plugin__
Export-Package: org.antlr.v4.runtime;uses:="org.antlr.v4.runtime.dfa,org
 .antlr.v4.runtime.atn,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tr
 ee,org.antlr.v4.runtime.tree.pattern,org.antlr.v4.runtime.tree.gui,java
 x.print,javax.swing";version="4.4.0.SNAPSHOT",org.antlr.v4.runtime.tree
 .gui;uses:="org.antlr.v4.runtime.tree,org.abego.treelayout,org.antlr.v4
 .runtime.misc,org.abego.treelayout.util,javax.swing,javax.swing.event,j
 avax.swing.tree,javax.swing.filechooser,javax.swing.border,javax.print,
 javax.imageio";version="4.4.0.SNAPSHOT",org.antlr.v4.runtime.tree.xpath
 ;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.tree,org.antlr.v4.run
 time.dfa,org.antlr.v4.runtime.atn";version="4.4.0.SNAPSHOT",org.antlr.v
 4.runtime.dfa;uses:="org.antlr.v4.runtime.atn,org.antlr.v4.runtime.misc
 ";version="4.4.0.SNAPSHOT",org.antlr.v4.runtime.atn;uses:="org.antlr.v4
 .runtime,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.dfa";version="4
 .4.0.SNAPSHOT",org.antlr.v4.runtime.tree.pattern;uses:="org.antlr.v4.ru
 ntime.misc,org.antlr.v4.runtime.tree,org.antlr.v4.runtime,org.antlr.v4.
 runtime.tree.xpath,org.antlr.v4.runtime.atn";version="4.4.0.SNAPSHOT",o
 rg.antlr.v4.runtime.misc;uses:="javax.swing,javax.imageio,javax.print,j
 avax.print.attribute,org.antlr.v4.runtime,org.antlr.v4.runtime.atn";ver
 sion="4.4.0.SNAPSHOT",org.antlr.v4.runtime.tree;uses:="org.antlr.v4.run
 time,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tree.gui";version="
 4.4.0.SNAPSHOT"
__Implementation-Title: ANTLR 4 Runtime__
__Implementation-Vendor: ANTLR__
__Implementation-Vendor-Id: org.antlr__
__Implementation-Version: 4.4-SNAPSHOT__
Import-Package: javax.imageio,javax.print,javax.print.attribute,javax.sw
 ing,javax.swing.border,javax.swing.event,javax.swing.filechooser,javax.
 swing.tree,org.abego.treelayout,org.abego.treelayout.util
Tool: Bnd-1.50.0

and secondly antlr4-annotations:

__Manifest-Version: 1.0__
Bnd-LastModified: 1409178802662
__Build-Jdk: 1.7.0_25__
__Built-By: cd01__
Bundle-Description: A set of annotations used within the ANTLR 4 Runtime
Bundle-DocURL: http://www.antlr.org
Bundle-License: http://www.antlr.org/license.html
Bundle-ManifestVersion: 2
Bundle-Name: ANTLR 4 Runtime Annotations
Bundle-SymbolicName: org.antlr.antlr4-annotations
Bundle-Vendor: ANTLR
Bundle-Version: 4.4.0.SNAPSHOT
__Created-By: Apache Maven Bundle Plugin__
Export-Package: org.antlr.v4.runtime.misc;uses:="javax.lang.model.elemen
 t,javax.lang.model.util,javax.tools,javax.lang.model,javax.lang.model.t
 ype,javax.annotation.processing";version="4.4.0.SNAPSHOT"
Fragment-Host: org.antlr.antlr4-runtime
__Implementation-Title: ANTLR 4 Runtime Annotations__
__Implementation-Vendor: ANTLR__
__Implementation-Vendor-Id: org.antlr__
__Implementation-Version: 4.4-SNAPSHOT__
Import-Package: javax.annotation.processing,javax.lang.model,javax.lang.
 model.element,javax.lang.model.type,javax.lang.model.util,javax.tools
Tool: Bnd-1.50.0

*ok, there is a slight difference in that with maven jar plugin you will get **Archiver-Version: Plexus Archiver*** :wink:

### CODE REVIEW

**Comment by parrt** [comment]
Ok, i'm closing as I automated this into bild.py script. runtime jar has osgi cruft.

**Comment by Samoht-fr** [comment]
Thanks

File: pom.xml [hunk 1]
```diff
@@ -75,6 +75,24 @@
         <bootclasspath.compile>${bootclasspath.java6}</bootclasspath.compile>
         <bootclasspath.testCompile>${bootclasspath.java6}</bootclasspath.testCompile>
         <antlr.testinprocess>true</antlr.testinprocess>
+        <maven-bundle-plugin.version>2.3.7</maven-bundle-plugin.version>
+
+        <!-- Default OSGi metadata - see maven-bundle-plugin configuration -->
+        <osgi.symbolicName>${project.groupId}.${project.artifactId}</osgi.symbolicName>
+        <osgi.failok>false</osgi.failok>
+        <osgi.nouses>false</osgi.nouses>
+        <osgi.versionPolicy>${range;[===,=+);${@}}</osgi.versionPolicy>
```
Comment (deklanowski) [inline pom.xml]: Hi Sam, first off I apologise that it has taken this long to reply, I thought this was dead and wasn't paying attention to any update until Terence emailed me directly today. In the version policy the '@' macro means that the version of a package as it is provided in a Maven defined dependency in the build classpath, this is an argument to the range macro. BND macro syntax is a little cryptic. I believe that the use of the '@' here may be redundant but that is the way I have been using it and it has the desired effect.

File: pom.xml [hunk 2]
```diff
@@ -75,6 +75,24 @@
...
+        <osgi.import>
+            *
+        </osgi.import>
+        <osgi.export/>
+        <!--
+            antlr4-runtime is main bundle and host to annotations fragment bundle
+            so define symbolic name here to ensure consistent reference in annotation
+            OSGi metadata.
+        -->
+        <antlr4-runtime.symbolicName>${project.groupId}.antlr4-runtime</antlr4-runtime.symbolicName>
```
Comment (deklanowski) [inline pom.xml]: Sure thing.

File: pom.xml [hunk 3]
```diff
@@ -75,6 +75,24 @@
         <bootclasspath.compile>${bootclasspath.java6}</bootclasspath.compile>
         <bootclasspath.testCompile>${bootclasspath.java6}</bootclasspath.testCompile>
         <antlr.testinprocess>true</antlr.testinprocess>
+        <maven-bundle-plugin.version>2.3.7</maven-bundle-plugin.version>
```
Comment (deklanowski) [inline pom.xml]: Of course I can inline the version number and I will test with a more current version of the bundle plugin.

File: pom.xml [hunk 4]
```diff
@@ -135,6 +153,29 @@
...
+        <pluginManagement>
+            <plugins>
+                <plugin>
+                    <groupId>org.apache.felix</groupId>
+                    <artifactId>maven-bundle-plugin</artifactId>
+                    <version>${maven-bundle-plugin.version}</version>
+                    <extensions>true</extensions>
+                    <configuration>
+                        <obrRepository>NONE</obrRepository>
+                        <manifestLocation>${project.build.directory}/META-INF</manifestLocation>
```
Comment (deklanowski) [inline pom.xml]: No reason, I just copied a pattern I have been using for a while but its perfectly fine to leave at the default. I will modify this.

**Comment by deklanowski** [comment]
Main issue with this is that abego treelayout dependency is not OSGi ready. This is a google code project with a single committer. I put in a request to update this project with OSGI metadata but received no response. Here is relevant email https://code.google.com/p/treelayout/issues/detail?id=2. The only other option is to embed this directly in the antlr4 runtime bundle classpath, this could be done for now and once it is properly OSGi-ified we can un-embed this. As I understand this artifact is only concerned with displaying tree structures of grammar and parse output. Please advise.

**Comment by deklanowski** [comment]
One other option is to make the resolution of abego.treelayout packages optional this assumes that one possibly might not use treelayout in an OSGi context. This means that the antlr4 runtime bundle will still start even with the treelayout bundle not present, however any usage of it will result in an Exception. The embedded option is probably safest and doesn't really add much bloat to the runtime jar (treelayout jat is only 25KB)

**Comment by parrt** [comment]
Howdy!  That tree layout stuff is specifically designed for ANTLR or at least I should say I commissioned it so we are free to embedded directly, as the current jar does. Shouldn't that relieve any issues with dependencies? The bnd tool gives the following output for the runtime jar:

```
terence:master:~/antlr/code/antlr4/dist $ bnd antlr4-4.5.jar 
[MANIFEST antlr4-4.5]
Bnd-LastModified                         1421374406005                           
Build-Jdk                                1.7.0_25                                
Built-By                                 parrt                                   
Bundle-Description                       The ANTLR 4 Runtime                     
Bundle-DocURL                            http://www.antlr.org                    
Bundle-License                           http://www.antlr.org/license.html       
Bundle-ManifestVersion                   2                                       
Bundle-Name                              ANTLR 4 Runtime                         
Bundle-SymbolicName                      org.antlr.antlr4-runtime                
Bundle-Vendor                            ANTLR                                   
Bundle-Version                           4.5                                     
Created-By                               1.7.0_25 (Oracle Corporation)           
Export-Package                           org.abego.treelayout,org.abego.treelayout.internal.util,org.abego.treelayout.internal.util.java.lang,org.abego.treelayout.internal.util.java.lang.string,org.abego.treelayout.internal.util.java.util,org.abego.treelayout.util;uses:="org.abego.treelayout",org.antlr.v4.runtime;uses:="javax.print,javax.swing,org.antlr.v4.runtime.atn,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tree,org.antlr.v4.runtime.tree.pattern",org.antlr.v4.runtime.atn;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.misc",org.antlr.v4.runtime.dfa;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.atn",org.antlr.v4.runtime.misc;uses:="javax.annotation.processing,javax.lang.model,javax.lang.model.element,javax.print,javax.swing,org.antlr.v4.runtime",org.antlr.v4.runtime.tree;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.misc",org.antlr.v4.runtime.tree.gui;uses:="javax.print,javax.swing,org.abego.treelayout,org.antlr.v4.runtime.tree",org.antlr.v4.runtime.tree.pattern;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tree",org.antlr.v4.runtime.tree.xpath;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.atn,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.tree"
Implementation-Title                     ANTLR 4 Runtime                         
Implementation-Vendor                    ANTLR                                   
Implementation-Vendor-Id                 org.antlr                               
Implementation-Version                   4.5                                     
Import-Package                           javax.annotation.processing;resolution:=optional,javax.imageio;resolution:=optional,javax.lang.model;resolution:=optional,javax.lang.model.element;resolution:=optional,javax.lang.model.type;resolution:=optional,javax.lang.model.util;resolution:=optional,javax.print;resolution:=optional,javax.print.attribute;resolution:=optional,javax.swing;resolution:=optional,javax.swing.border;resolution:=optional,javax.swing.event;resolution:=optional,javax.swing.filechooser;resolution:=optional,javax.swing.tree;resolution:=optional,javax.tools;resolution:=optional
Manifest-Version                         1.0                                     
Originally-Created-By                    http://www.bildtool.org                 
Require-Capability                       osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.6))"
Tool                                     Bnd-2.4.0.201411031536                  
```

It looks like it is exporting not trying to import the tree layout stuff. Perhaps you could just download it and give it a try? http://www.antlr.org/download/antlr-runtime-4.5-rc-2-osgi.jar

**Comment by deklanowski** [comment]
Hi Terence, if you use BND without giving it specific direction then it will try to infer what should be exported and imported by looking at the classpath, while we can do that by supplying it instructions in a supplementary configuration the maven-bundle-plugin effectively helps us achieve that within the project pom. Once Udo applies my suggested config for treelayout we can proceed that way. If you also want to be able to do something similar with your python build script, I can replicate the config so you can do a command line build with BND which achieves the same result.

**Comment by parrt** [comment]
It would be great to have the necessary commandline configurations, thanks! I like to stick with building this with my Python script. We won't be building the software with maven so I don't think there's a point in updating that. Thanks for all your help and patience! I think we're almost there...

**Comment by deklanowski** [comment]
I verified Udo's release 1.0.2 and it looks good. I need a bit more time to translate the config necessary for command line BND, bear with me, shouldn't take to long. BTW I find maven a necessary evil, personally I think it's a PITA, I fully understand wanting to regain control of the build, would love to hear of your experiences sometime.

**Comment by parrt** [comment]
Cool. but we're still ok embedding as I've done, right, in the runtime?

**Comment by deklanowski** [comment]
We can do that but we dont need to, the whole point of making Udo's treelayout OSGi ready is that it can be installed in OSGi container and export the required packages for the Antlr 4 runtime, it leaves users free to use treelayout in it's own right. You did mention that you commissioned it for Antlr which makes this point moot I suppose. Anyway, I'll give you the first decoupled option, the second embedded option is pretty simple too, I will add that one probably tomorrow as its getting a bit late here. 
This BND file simply exports the Antlr 4 runtime packages and imports treelayout packages, it expects to find the treelayout bundle in the same OSGi container. It's real simple, here is the file - you can simply update versions etc as appropriate:

`#-----------------------------------------------------------------
# BND instructions to make antlr4-runtime OSGi-ready
#
# Import packages will be inferred by inspection. See generated 
# MANIFEST.MF
#-----------------------------------------------------------------
Bundle-Name: ANTLR 4 Runtime 
Bundle-Description: The ANTLR 4 Runtime 
Bundle-DocURL: http://www.antlr.org
Bundle-SymbolicName: org.antlr.antlr4-runtime
Bundle-Version: 4.5

# Export all runtime packages
Export-Package: org.antlr.v4.runtime.*;version="4.5"

# end of file`

Just name this org.antlr.antlr4-runtime.bnd. Here is how I ran it:

`$ java -jar bnd-2.4.0.jar bnd buildx -c antlr4-runtime-4.4-SNAPSHOT.jar org.antlr.antlr4-runtime.bnd`

Things to note:
The SNAPSHOT jar is a vanilla jar from a java build, the output jar file prefix will be the bnd file name with the addition of a version If you want to override this add a -o <jar file name>. The MANIFEST produced is as follows:
`Manifest-Version: 1.0
Bnd-LastModified: 1421709307443
Bundle-Description: The ANTLR 4 Runtime 
Bundle-DocURL: http://www.antlr.org
Bundle-ManifestVersion: 2
Bundle-Name: ANTLR 4 Runtime 
Bundle-SymbolicName: org.antlr.antlr4-runtime
Bundle-Version: 4.5
Created-By: 1.7.0_45 (Oracle Corporation)
Export-Package: org.antlr.v4.runtime;version="4.5";uses:="javax.print,ja
 vax.swing,org.antlr.v4.runtime.atn,org.antlr.v4.runtime.dfa,org.antlr.v
 4.runtime.misc,org.antlr.v4.runtime.tree,org.antlr.v4.runtime.tree.patt
 ern",org.antlr.v4.runtime.atn;version="4.5";uses:="org.antlr.v4.runtime
 ,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.misc",org.antlr.v4.runti
 me.dfa;version="4.5";uses:="org.antlr.v4.runtime.atn",org.antlr.v4.runt
 ime.misc;version="4.5";uses:="javax.print,javax.swing,org.antlr.v4.runt
 ime",org.antlr.v4.runtime.tree;version="4.5";uses:="org.antlr.v4.runtim
 e,org.antlr.v4.runtime.misc",org.antlr.v4.runtime.tree.gui;version="4.5
 ";uses:="javax.print,javax.swing,org.abego.treelayout,org.antlr.v4.runt
 ime.tree",org.antlr.v4.runtime.tree.pattern;version="4.5";uses:="org.an
 tlr.v4.runtime,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tree",org
 .antlr.v4.runtime.tree.xpath;version="4.5";uses:="org.antlr.v4.runtime,
 org.antlr.v4.runtime.atn,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.
 tree"
Import-Package: javax.imageio,javax.print,javax.print.attribute,javax.sw
 ing,javax.swing.border,javax.swing.event,javax.swing.filechooser,javax.
 swing.tree,org.abego.treelayout,org.abego.treelayout.util
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.6))"
Tool: Bnd-2.4.0.201411031536`

My tests resolve just fine with Udo's 1.0.2 treelayout version which is now available in maven central, this would keep all the OSGi folks happy.

**Comment by parrt** [comment]
ok thanks for the help.

On Mon, Jan 19, 2015 at 3:21 PM, Declan Cox <notifications@github.com>
wrote:

> We can do that but we dont need to, the whole point of making Udo's
> treelayout OSGi ready is that it can be installed in OSGi container and
> export the required packages for the Antlr 4 runtime, it leaves users free
> to use treelayout in it's own right. You did mention that you commissioned
> it for Antlr which makes this point moot I suppose. Anyway, I'll give you
> the first decoupled option, the second embedded option is pretty simple
> too, I will add that one probably tomorrow as its getting a bit late here.
> This BND file simply exports the Antlr 4 runtime packages and imports
> treelayout packages, it expects to find the treelayout bundle in the same
> OSGi container. It's real simple, here is the file - you can simply update
> versions etc as appropriate:
>
> `#-----------------------------------------------------------------
> BND instructions to make antlr4-runtime OSGi-ready
>
> #
> Import packages will be inferred by inspection. See generated MANIFEST.MF
>
> #-----------------------------------------------------------------
> Bundle-Name: ANTLR 4 Runtime
> Bundle-Description: The ANTLR 4 Runtime
> Bundle-DocURL: http://www.antlr.org
> Bundle-SymbolicName: org.antlr.antlr4-runtime
> Bundle-Version: 4.5
> Export all runtime packages
>
> Export-Package: org.antlr.v4.runtime.*;version="4.5"
> end of file`
>
> Just name this org.antlr.antlr4-runtime.bnd. Here is how I ran it:
>
> $ java -jar bnd-2.4.0.jar bnd buildx -c antlr4-runtime-4.4-SNAPSHOT.jar
> org.antlr.antlr4-runtime.bnd
>
> Things to note:
> The SNAPSHOT jar is a vanilla jar from a java build, the output jar file
> prefix will be the bnd file name with the addition of a version If you want
> to override this add a -o . The MANIFEST produced is as follows:
> Manifest-Version: 1.0
> Bnd-LastModified: 1421709307443
> Bundle-Description: The ANTLR 4 Runtime
> Bundle-DocURL: http://www.antlr.org
> Bundle-ManifestVersion: 2
> Bundle-Name: ANTLR 4 Runtime
> Bundle-SymbolicName: org.antlr.antlr4-runtime
> Bundle-Version: 4.5
> Created-By: 1.7.0_45 (Oracle Corporation)
> Export-Package: org.antlr.v4.runtime;version="4.5";uses:="javax.print,ja
> vax.swing,org.antlr.v4.runtime.atn,org.antlr.v4.runtime.dfa,org.antlr.v
> 4.runtime.misc,org.antlr.v4.runtime.tree,org.antlr.v4.runtime.tree.patt
> ern",org.antlr.v4.runtime.atn;version="4.5";uses:="org.antlr.v4.runtime
> ,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.misc",org.antlr.v4.runti
> me.dfa;version="4.5";uses:="org.antlr.v4.runtime.atn",org.antlr.v4.runt
> ime.misc;version="4.5";uses:="javax.print,javax.swing,org.antlr.v4.runt
> ime",org.antlr.v4.runtime.tree;version="4.5";uses:="org.antlr.v4.runtim
> e,org.antlr.v4.runtime.misc",org.antlr.v4.runtime.tree.gui;version="4.5
> ";uses:="javax.print,javax.swing,org.abego.treelayout,org.antlr.v4.runt
> ime.tree",org.antlr.v4.runtime.tree.pattern;version="4.5";uses:="org.an
> tlr.v4.runtime,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tree",org
> .antlr.v4.runtime.tree.xpath;version="4.5";uses:="org.antlr.v4.runtime,
> org.antlr.v4.runtime.atn,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.
> tree"
> Import-Package: javax.imageio,javax.print,javax.print.attribute,javax.sw
> ing,javax.swing.border,javax.swing.event,javax.swing.filechooser,javax.
> swing.tree,org.abego.treelayout,org.abego.treelayout.util
> Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.6))"
> Tool: Bnd-2.4.0.201411031536
>
> My tests resolve just fine with Udo's 1.0.2 treelayout version which is
> now available in maven central, this would keep all the OSGi folks happy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/antlr/antlr4/pull/689#issuecomment-70578638>.
>



-- 
Dictation in use. Please excuse homophones, malapropisms, and nonsense.

**Comment by deklanowski** [comment]
No problem, sorry about the markup screwups there. Let me know if I can be of more help.

**Pull request closed without merging**